lib/src/streams/stream_handler.dart

```diff
@@ -330,6 +330,13 @@ class StreamHandler extends Object with TerminatableMixin, ClosableMixin {
 
     _setupOutgoingMessageHandling(stream);
 
+    // handle incoming stream cancelation
+    streamQueueIn.onCancel.then((_) {
+      if (!streamQueueIn.wasClosed && !streamQueueIn.wasTerminated) {
+        _frameWriter.writeRstStreamFrame(streamId, ErrorCode.CANCEL);
+      }
```

Comment:
After reading the http/2 spec for some time this morning, it seems to me that this is not quite the right thing to do.

Let's assume we have a bi-directional stream between A and B. B now cancels the `incomingMessages` stream. Though B might still want to send more data to `A` via the `outgoingMessages` stream.

Though if `B` sends an RST.Cancel, it will cause `A` to consider the stream as closed in both directions, see the [http/2 state diagram](https://http2.github.io/http2-spec/#rfc.section.5.1), which is not what we intended.

So far I have not found any mechanism in http/2 to tell the remote end we are not interested in more data anymore without also saying that we are not interested in sending data anymore. Only the sender can cause the state to go into half-closed by sending an end-of-stream.

Maybe the best thing to do is simply drop incoming messages from the remote end and keep everything else as it is. If we end up in the "half-closed (local)" state **and** we cancelled incoming messages it is safe to send the RST.

(We could argue that canceling for `incomingMessages` is the same as terminating the stream (in which case sending an RST.Cancel frame is ok) - but this seems not quite right to me from an API point of view - a user would not expect that. We do have a `stream.terminate()` method this after all)